source/heating_model/latent_heat_melt.cc

```diff
@@ -36,25 +36,60 @@ namespace aspect
       Assert(heating_model_outputs.heating_source_terms.size() == material_model_inputs.position.size(),
              ExcMessage ("Heating outputs need to have the same number of entries as the material model inputs."));
 
+      const bool use_operator_split = (this->get_parameters().use_operator_splitting);
+
+      const MaterialModel::ReactionRateOutputs<dim> *reaction_rate_out
+        = material_model_outputs.template get_additional_output<MaterialModel::ReactionRateOutputs<dim> >();
+
       for (unsigned int q=0; q<heating_model_outputs.heating_source_terms.size(); ++q)
         {
+          heating_model_outputs.heating_source_terms[q] = 0.0;
+          heating_model_outputs.lhs_latent_heat_terms[q] = 0.0;
+          heating_model_outputs.rates_of_temperature_change[q] = 0.0;
+
           if (this->introspection().compositional_name_exists("porosity") &&  this->get_timestep_number() > 0)
             {
-              // with melt migration the reaction term is a mass reaction rate
               const double porosity_idx = this->introspection().compositional_index_for_name("porosity");
-              const double latent_heat = melting_entropy_change * material_model_outputs.reaction_terms[q][porosity_idx];
-              heating_model_outputs.heating_source_terms[q] = latent_heat
-                                                              * material_model_inputs.temperature[q];
-
-              // without melt migration, the reaction term is a constant value in terms of volume,
-              // and we have to scale it to the correct units
-              if (!this->include_melt_transport())
-                heating_model_outputs.heating_source_terms[q] *= material_model_outputs.densities[q] / this->get_timestep();
+              double melting_rate = 0.0;
+
+              if (!use_operator_split)
+                {
+                  // with melt migration the reaction term is a mass reaction rate,
+                  // without melt migration, the reaction term is a constant value in terms of volume,
+                  // and we have to scale it to the correct units
+                  if (this->include_melt_transport())
+                    melting_rate = material_model_outputs.reaction_terms[q][porosity_idx];
+                  else
+                    melting_rate = material_model_outputs.reaction_terms[q][porosity_idx]
+                                   * material_model_outputs.densities[q] / this->get_timestep();
+
+                  heating_model_outputs.heating_source_terms[q] = melting_entropy_change
+                                                                  * melting_rate
+                                                                  * material_model_inputs.temperature[q];
+                }
+              else if (use_operator_split && reaction_rate_out != NULL)
```

Comment:
But I think there are many places in the code where we evaluate the heating model, but do not call `create_additional_named_outputs` before, and in those cases reaction_rate_out == NULL (like in the normal assembly). That is a totally valid case. The entries in reaction_rate_out just have to be filled in case it exists, so they can't be NaNs (which is covered in the assert below), but if the `ReactionRateOutputs` have not been created, that probably just means that in the place where we are in the code they are not needed.